plugins/modules/foreman_host.py

```diff
@@ -86,76 +119,62 @@
 
 RETURN = ''' # '''
 
-try:
-    from nailgun.entities import (
-        Host,
-        HostGroup,
-    )
-
-    from ansible.module_utils.ansible_nailgun_cement import (
-        find_host,
-        find_hostgroup,
-        find_location,
-        find_organization,
-        ForemanEntityAnsibleModule,
-        naildown_entity_state,
-        sanitize_entity_dict,
-    )
-except ImportError:
-    pass
-
-
-# This is the only true source for names (and conversions thereof)
-name_map = {
-    'name': 'name',
-    'enabled': 'enabled',
-    'managed': 'managed',
-    'hostgroup': 'hostgroup',
-    'location': 'location',
-    'organization': 'organization',
-}
+from ansible.module_utils.foreman_helper import ForemanEntityAnsibleModule
 
 
 def main():
     module = ForemanEntityAnsibleModule(
-        argument_spec=dict(
+        entity_spec=dict(
             name=dict(required=True),
-            hostgroup=dict(),
-            location=dict(),
-            organization=dict(),
-            enabled=dict(default='true', type='bool'),
+            hostgroup=dict(type='entity', flat_name='hostgroup_id'),
+            location=dict(type='entity', flat_name='location_id'),
+            organization=dict(type='entity', flat_name='organization_id'),
+            enabled=dict(type='bool'),
             managed=dict(type='bool'),
-            state=dict(default='present', choices=[
-                       'present_with_defaults', 'present', 'absent']),
+            build=dict(type='bool'),
         ),
         required_if=(
-            ['state', 'present_with_defaults', ['hostgroup']],
-            ['state', 'present', ['hostgroup']],
+            ['managed', True, ['hostgroup']],
+            ['build', True, ['hostgroup']],
         ),
     )
 
-    (host_dict, state) = module.parse_params()
+    entity_dict = module.clean_params()
+
+    # additional param validation
+    if not module.desired_absent:
+        if 'hostgroup' not in entity_dict and entity_dict.get('managed', True):
+            module.fail_json(msg='Hostgroup can be omitted only with managed=False')
+
+        if 'build' in entity_dict:
+            # When 'build'=True, 'managed' has to be True.
+            # Assuming that user's priority is to build
```

Comment:
I tried a few different approaches. This one seems most logical to me.
There are two contradicting facts:
* when ```build```=true, managed has to be true, otherwise call fails
* when ```managed=false``` (we want to unmanage host) and existing host has both at true, we have to pass ```build=false```, otherwise call fails
So I decided that if user passes ```build``` explicitly then it takes precedence.

Reply:
I think that totally makes sense. Just wondering if the following code would be more readable:

```python
if 'build' in entity_dict and entity_dict['build'] and 'managed' in entity_dict and not entity_dict['managed']:
     # When 'build'=True, 'managed' has to be True. Assuming that user's priority is to build.
      module.warn("when 'build'=True, 'managed' is ignored and forced to True")
      entity_dict['managed'] = True
elif 'build' not in entity_dict and 'managed' in entity_dict and not entity_dict['managed']:
    # When 'build' is not given and 'managed'=False, have to clear 'build' context that might exist on the server
    entity_dict['build'] = False
```

Reply:
Sure, will collapse the conditions.
This one above will allow for managed=false build=true if managed omitted and host exists with managed=false. Will fail with "can't build unmanaged host"